### PR TITLE
Url encode password

### PR DESCRIPTION
**Pull request by m9tdev**

### CODE REVIEW

File: arris_tg2492lg/connect_box.py [hunk 1]
```diff
@@ -30,16 +31,15 @@ class ConnectBox:
...
         self._nonce = str(random.randrange(10000, 100000))
         self._credential: Optional[Credential] = None
 
     async def async_login(self) -> str:
         arg_string = f"{USERNAME}:{self._password}"
         arg = base64.b64encode(arg_string.encode("utf-8")).decode("ascii")
 
-        params = {"arg": arg, "_n": self._nonce}
-        async with self._websession.get(f"{self._hostname}/login", params=params) as response:
+        async with self._websession.get(f"{self._hostname}/login?arg={arg}&_n={self._nonce}") as response:
```
Comment (vanbalken) [inline arris_tg2492lg/connect_box.py:42]: What is your motivation to move the query parameters from `params` to the url?

**Comment by vanbalken** [comment]
@mathieupost Thanks for creating this pull request! This is valid issue. I compared your change with the javascript code used by the webpage of the router to make sure they do the same. The webpage also escapes special characters both for password and the username.

Do you use this library in your own code or as part of Home Assistant? If you use it with Home Assistant this will also need to be fixed on the 1.x version of this library and Home Assistant Core will need to be updated.

**Comment by m9tdev** [comment]
My pleasure!

I was planning on using it in Home Assistant, but I was just trying out the library locally and found out that my password gave a 401. So I started debugging and found out that this was the issue ;)

I actually didn't know that there was integration with Home Assistant already! Can you point me to where I can find it? I'll look into fixing it there as well.

**Comment by vanbalken** [comment]
> My pleasure!
> 
> I was planning on using it in Home Assistant, but I was just trying out the library locally and found out that my password gave a 401. So I started debugging and found out that this was the issue ;)
> 
> I actually didn't know that there was integration with Home Assistant already! Can you point me to where I can find it? I'll look into fixing it there as well.

It is used in this integration: https://www.home-assistant.io/integrations/arris_tg2492lg/ I was planning on updating the integration so it can be configured in the UI but personal circumstances leave me with little time for that.

A quick fix would be to manually escape the special characters and than you can use the integration right away.

I will try to apply this fix to the 1.x version this weekend. Then we only need to bump the version of this library used in Home Assistant.

File: arris_tg2492lg/connect_box.py (hunk 1 above)
Comment (m9tdev) [reply arris_tg2492lg/connect_box.py:42]: For some reason the special characters in the base64 encoded string in the 'arg' param were encoded again (e.g. the trailing '=' was encoded as '%3D'), which would result in an internal server error from the router.

**Comment by m9tdev** [comment]
Cool! I'll use the quick fix in the mean time. I'll look into it when I have time

File: arris_tg2492lg/connect_box.py [hunk 2]
```diff
@@ -30,16 +31,15 @@ class ConnectBox:
     def __init__(self, websession: ClientSession, hostname: str, password: str):
         self._websession = websession
         self._hostname = hostname
-        self._password = password
+        self._password = quote(password)
```
Comment (vanbalken) [inline arris_tg2492lg/connect_box.py:34]: I would call `quote()` when creating the `arg_string` so the password stays unchanged here.

File: arris_tg2492lg/connect_box.py [hunk 3]
```diff
@@ -30,16 +31,15 @@ class ConnectBox:
     def __init__(self, websession: ClientSession, hostname: str, password: str):
         self._websession = websession
         self._hostname = hostname
-        self._password = password
+        self._password = quote(password)
         self._nonce = str(random.randrange(10000, 100000))
         self._credential: Optional[Credential] = None
 
     async def async_login(self) -> str:
         arg_string = f"{USERNAME}:{self._password}"
```
Comment (vanbalken) [inline arris_tg2492lg/connect_box.py:39]: `arg_string = f"{quote(USERNAME)}:{quote(self._password)}"`

Although `USERNAME` (admin) doesn't need the be escaped I prefer to keep this in line with how the router portal works. If the username would ever be configurable I won't forget to escape it :)

File: arris_tg2492lg/connect_box.py (hunk 1 above)
Comment (vanbalken) [reply arris_tg2492lg/connect_box.py:42]:
```
params = f"arg={arg}&_n={self.nonce}"
async with self._websession.get(f"{self._hostname}/login", params=params) as response:
```
It's a bit personal preference but if this works I would prefer it this way.

**Review by vanbalken — CHANGES_REQUESTED** [review]
I fixed this for the 1.x version of this library which is used by Home Assisant: https://github.com/vanbalken/arris-tg2492lg/pull/29 The new version has been published on [PyPI](https://pypi.org/project/arris-tg2492lg/1.2.1/) and I only need to create a pullrequest to Home Assistant. I will try to do that this weekend.

I made some small changes in de the 1.x fix compared to this PR. If you could apply those here that would be great! I will add some unit tests later on.

File: arris_tg2492lg/connect_box.py (hunk 1 above)
Comment (vanbalken) [reply arris_tg2492lg/connect_box.py:42]: The params are encoded by aiohttp when passed separately.

**Comment by vanbalken** [comment]
Close pull request as changes have been incorporated in https://github.com/vanbalken/arris-tg2492lg/pull/41 which also includes unit tests.

Exactly 2 years later (what a coincidence :) ) and finally done :sweat_smile:

**Pull request closed without merging**